pstree issues: Consistent post-process fails due to hard-coded Mocked parent PID

Just a heads-up.

Well, I might've been lucky here... I've installed the sandbox quite a few times in the past year, and everything has always worked fine... Until this last installation, where the `drak-postprocess` consistently failed creating a process tree because of a `MultipleProcessesReturned` error. Here's a sample journald log:

```
Jan 25 09:47:55 DRAK-SBX-5 drak-postprocess[671]: [2022-01-25 09:47:55,813][INFO] Received new task - 00ffc247-4e76-4966-a562-80c999f16021
Jan 25 09:47:55 DRAK-SBX-5 drak-postprocess[671]: INFO:karton.drakrun.processor:Received new task - 00ffc247-4e76-4966-a562-80c999f16021
Jan 25 09:48:37 DRAK-SBX-5 drak-postprocess[671]: ERROR:root:Failed to parse 49 lines generated by unknown
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]: WARNING:root:Failed to process {'Plugin': 'procmon', 'TimeStamp': '1643099487.878433', 'PID': 336, 'PPID': 436, 'RunningPro>
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]: [2022-01-25 09:48:46,425][ERROR] Postprocess failed
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]: Traceback (most recent call last):
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/process.py", line 98, in process
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:     outputs = plugin.handler(
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 241, in build_process_tree
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:     data = json.dumps(tree_from_log(tmp_file)).encode()
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 235, in tree_from_log
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:     raise e
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 220, in tree_from_log
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:     parse_running_process_entry(pstree, entry)
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 127, in parse_running_process>
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:     pstree.add_process(p)
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 38, in add_process
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:     raise MultipleProcessesReturned(processes)
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]: drakcore.postprocess.pstree.MultipleProcessesReturned: More than one proces fulfills condition: Mocked parent(336) []
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]: ERROR:karton.drakrun.processor:Postprocess failed
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]: Traceback (most recent call last):
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/process.py", line 98, in process
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:     outputs = plugin.handler(
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 241, in build_process_tree
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:     data = json.dumps(tree_from_log(tmp_file)).encode()
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 235, in tree_from_log
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:     raise e
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 220, in tree_from_log
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:     parse_running_process_entry(pstree, entry)
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 127, in parse_running_process>
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:     pstree.add_process(p)
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 38, in add_process
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]:     raise MultipleProcessesReturned(processes)
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]: drakcore.postprocess.pstree.MultipleProcessesReturned: More than one proces fulfills condition: Mocked parent(336) []
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]: [2022-01-25 09:48:46,427][INFO] Skipping generate_wireshark_key_file, missing resources
Jan 25 09:48:46 DRAK-SBX-5 drak-postprocess[671]: INFO:karton.drakrun.processor:Skipping generate_wireshark_key_file, missing resources
Jan 25 09:48:50 DRAK-SBX-5 drak-postprocess[671]: [2022-01-25 09:48:50,913][INFO] Task done - 00ffc247-4e76-4966-a562-80c999f16021
Jan 25 09:48:50 DRAK-SBX-5 drak-postprocess[671]: INFO:karton.drakrun.processor:Task done - 00ffc247-4e76-4966-a562-80c999f16021
Jan 25 09:48:51 DRAK-SBX-5 drak-postprocess[671]: [2022-01-25 09:48:50,916][INFO] Received new task - 5a8f6ae6-6b06-4e0a-98ee-d5333201f5a8
Jan 25 09:48:51 DRAK-SBX-5 drak-postprocess[671]: INFO:karton.drakrun.processor:Received new task - 5a8f6ae6-6b06-4e0a-98ee-d5333201f5a8
Jan 25 09:49:45 DRAK-SBX-5 drak-postprocess[671]: ERROR:root:Failed to parse 49 lines generated by unknown
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]: WARNING:root:Failed to process {'Plugin': 'procmon', 'TimeStamp': '1643099487.705992', 'PID': 336, 'PPID': 436, 'RunningPro>
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]: [2022-01-25 09:49:53,876][ERROR] Postprocess failed
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]: Traceback (most recent call last):
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/process.py", line 98, in process
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:     outputs = plugin.handler(
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 241, in build_process_tree
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:     data = json.dumps(tree_from_log(tmp_file)).encode()
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 235, in tree_from_log
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:     raise e
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 220, in tree_from_log
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:     parse_running_process_entry(pstree, entry)
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 127, in parse_running_process>
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:     pstree.add_process(p)
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 38, in add_process
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:     raise MultipleProcessesReturned(processes)
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]: drakcore.postprocess.pstree.MultipleProcessesReturned: More than one proces fulfills condition: Mocked parent(336) []
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]: ERROR:karton.drakrun.processor:Postprocess failed
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]: Traceback (most recent call last):
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/process.py", line 98, in process
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:     outputs = plugin.handler(
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 241, in build_process_tree
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:     data = json.dumps(tree_from_log(tmp_file)).encode()
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 235, in tree_from_log
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:     raise e
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 220, in tree_from_log
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:     parse_running_process_entry(pstree, entry)
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 127, in parse_running_process>
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:     pstree.add_process(p)
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:   File "/opt/venvs/drakcore/lib/python3.8/site-packages/drakcore/postprocess/pstree.py", line 38, in add_process
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]:     raise MultipleProcessesReturned(processes)
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]: drakcore.postprocess.pstree.MultipleProcessesReturned: More than one proces fulfills condition: Mocked parent(336) []
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]: [2022-01-25 09:49:53,878][INFO] Skipping generate_wireshark_key_file, missing resources
Jan 25 09:49:53 DRAK-SBX-5 drak-postprocess[671]: INFO:karton.drakrun.processor:Skipping generate_wireshark_key_file, missing resources
Jan 25 09:49:57 DRAK-SBX-5 drak-postprocess[671]: [2022-01-25 09:49:57,198][INFO] Task done - 5a8f6ae6-6b06-4e0a-98ee-d5333201f5a8
Jan 25 09:49:57 DRAK-SBX-5 drak-postprocess[671]: INFO:karton.drakrun.processor:Task done - 5a8f6ae6-6b06-4e0a-98ee-d5333201f5a8

```

Sure enough, spinning up the VM again to check running processes, there happened to be a process with PID 336, the same as the hard-coded one for mocking a parent process in the process-tree generation code :smile: 

![Screenshot from 2022-01-25 10-09-05](https://user-images.githubusercontent.com/12098890/150953571-db8a2921-3f52-44dd-bca1-4d660f469c81.png)

Not sure if there would be a way of assigning that mocked parent PID during the postinstall, checking that there's currently no other process with it?

Anyway, just wanted to let you know about this :man_shrugging:  I fixed by rebooting the VM and checking that no process had PID 336

--- a/drakrun/drakrun/lib/postprocessing/build_process_tree.py
+++ b/drakrun/drakrun/lib/postprocessing/build_process_tree.py
@@ -147,7 +147,7 @@ def parse_nt_create_user_process_entry(
     )
     if parent is None:
         # Parent must be alive at the process creation time, but who knows what happened
-        logger.debug(
+        logger.warning(
             f"Parent process not found at the process creation time (PID: {process_pid}, PPID: {process_ppid})"
         )
     else:
@@ -177,7 +177,7 @@ def parse_nt_create_process_ex_entry(
     )
     if parent is None:
         # Parent must be alive at the process creation time, but who knows what happened
-        logger.debug(
+        logger.warning(
             f"Parent process not found at the process creation time (PID: {process_pid}, PPID: {process_ppid})"
         )
     else:

--- a/drakrun/drakrun/lib/postprocessing/build_process_tree.py
+++ b/drakrun/drakrun/lib/postprocessing/build_process_tree.py
@@ -96,24 +96,10 @@ class MultipleProcessesReturned(Exception):
         super().__init__(message)
 
 
-class MissingParentProcessError(Exception):
-    def __init__(self, process: Process):
-        message = f"Cannot find parent process of {process}"
-        super().__init__(message)
-
-
 def parse_running_process_entry(pstree: ProcessTree, entry: Dict[str, Any]) -> None:
+    # We assume here that running processes are enumerated in order of creation
+    # (created by appending new entries to the end of EPROCESS linked list)
     parent = pstree.get_single_process(entry["PPID"], 0, float(entry["TimeStamp"]))
-    if parent is None:
-        # Running processes might have parents that we don't have any information about. Mock them.
-        parent = Process(
-            pid=entry["PPID"],
-            procname="Mocked parent",
-            ts_from=0.0,  # We don't know when the process was created.
-            # But we know it is not longer alive.
-            ts_to=float(entry["TimeStamp"]),
-        )
-        pstree.add_process(parent)
     p = Process(
         pid=entry["PID"],
         procname=entry["RunningProcess"],
@@ -145,11 +131,13 @@ def parse_nt_create_user_process_entry(
     if int(entry["Status"], 16) != 0:
         # Ignore unsuccessful entries.
         return
+    process_pid = entry["NewPid"]
+    process_ppid = entry["PID"]
     parent = pstree.get_single_process(
-        entry["PID"], float(entry["TimeStamp"]), float(entry["TimeStamp"])
+        process_ppid, float(entry["TimeStamp"]), float(entry["TimeStamp"])
     )
     p = Process(
-        pid=entry["NewPid"],
+        pid=process_pid,
         procname=entry["ImagePathName"],
         ts_from=float(entry["TimeStamp"]),
         # At this point, we don't know yet when the process will be terminated.
@@ -158,9 +146,12 @@ def parse_nt_create_user_process_entry(
         args=split_commandline(entry["CommandLine"]) if entry["CommandLine"] else [],
     )
     if parent is None:
-        # Parent must be alive at the process creation time.
-        raise MissingParentProcessError(p)
-    parent.children.append(p)
+        # Parent must be alive at the process creation time, but who knows what happened
+        logger.debug(
+            f"Parent process not found at the process creation time (PID: {process_pid}, PPID: {process_ppid})"
+        )
+    else:
+        parent.children.append(p)
     pstree.add_process(p)
 
 
@@ -171,11 +162,13 @@ def parse_nt_create_process_ex_entry(
     if int(entry["Status"], 16) != 0:
         # Ignore unsuccessful entries.
         return
+    process_pid = entry["NewPid"]
+    process_ppid = entry["PID"]
     parent = pstree.get_single_process(
-        entry["PID"], float(entry["TimeStamp"]), float(entry["TimeStamp"])
+        process_ppid, float(entry["TimeStamp"]), float(entry["TimeStamp"])
     )
     p = Process(
-        pid=entry["NewPid"],
+        pid=process_pid,
         procname="Unnamed",
         ts_from=float(entry["TimeStamp"]),
         # At this point, we don't know yet when the process will be terminated.
@@ -183,9 +176,12 @@ def parse_nt_create_process_ex_entry(
         parent=parent,
     )
     if parent is None:
-        # Parent must be alive at the process creation time.
-        raise MissingParentProcessError(p)
-    parent.children.append(p)
+        # Parent must be alive at the process creation time, but who knows what happened
+        logger.debug(
+            f"Parent process not found at the process creation time (PID: {process_pid}, PPID: {process_ppid})"
+        )
+    else:
+        parent.children.append(p)
     pstree.add_process(p)
 
 

--- a/drakrun/drakrun/lib/postprocessing/build_process_tree.py
+++ b/drakrun/drakrun/lib/postprocessing/build_process_tree.py
@@ -108,7 +108,8 @@ def parse_running_process_entry(pstree: ProcessTree, entry: Dict[str, Any]) -> N
         ts_to=None,
         parent=parent,
     )
-    parent.children.append(p)
+    if parent is not None:
+        parent.children.append(p)
     pstree.add_process(p)
 
 

